### PR TITLE
Fetch feeds concurrently

### PR DESCRIPTION
**Pull request by harlow**
@croaky a slightly different approach using a channel to collect the feed items and then a timeout to stop execution after 3 seconds.

I'd be curious if @bernerdschaefer has a more elegant solution to this?

* Add timeout so slow feeds don't timeout the execution of script

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -29,44 +29,52 @@ func rssHandler(rw http.ResponseWriter, r *http.Request) {
...
 	result, _ := master.ToAtom()
 	fmt.Fprintln(rw, result)
 }
 
-func fetch(feed sourceFeed, master *feeds.Feed) {
-	fetcher := rss.New(5, true, chanHandler, makeHandler(master, feed.name))
-	client := &http.Client{
-		Timeout: time.Second,
+func addItems(itemCh chan *feeds.Item, master *feeds.Feed) {
+	timeout := time.After(3 * time.Second)
```
Comment (croaky) [inline main.go]: Is this causing the code to run for exactly 3 seconds? It seems like it's pretty reliably in that range at localhost:8080. I'm seeing between 1-1.5s per load in the non-concurrent production code at rss.thoughtbot.com

Comment (bernerdschaefer) [inline main.go]: I believe that's indeed what will happen here. There's no termination case for when all of the `fetch`es have completed, so the `itemCh` stays open.

**Comment by bernerdschaefer** [comment]
This parallelizes the processing of individual feed items, which seems less important that parallelizing the fetching of the feeds themselves. Something in this direction might be more straightforward but still a good win:

```go
feedItems := make(chan []*rss.Item, len(urls))

for url := range urls {
  go fetch(url, feedItems)  // fetch sends its complete filtered list of items
}

deadline := time.After(3 * time.Second)
for i := 0; i < len(urls); i++ {
  select {
  case items := <-feedItems:
    // merge items into master feed
  case <-deadline:
    // abort
  }
}
```

File: main.go (hunk 1 above)
Comment (harlow) [inline main.go]: @croaky good catch. You're right.... without closing the `itemCh` this code always runs for 3 seconds!

**Comment by croaky** [comment]
I like that `deadline` naming.

**Comment by harlow** [comment]
Spaced on completing this one... looks like has been implemented by @croaky in https://github.com/thoughtbot/rss/commit/1214c8036ea24a37a9539da4b35d18fae0bc028e. Closing.

**Pull request closed without merging**